Feat/mini challenge 4

PedroEdu6786 opened the pull request:
### PR Summarize
- Add ThemeContentProvider for handling UI theme state
- Add VideosProvider for handling videos state
- Add test cases for hooks, state reducers
- Add homepage, header and previewRelated list components tests
- Add 2nd delivery improvements

### Test coverage
![image](https://user-images.githubusercontent.com/50753661/129453597-f15087c8-18a4-483b-8ee3-432d5cf20b08.png)

PedroEdu6786 committed:
:sparkles: feat: add new HomePage screen with dynamic video list and styles

Add styled-components and react-icons library, new header, layout, homepage and videopreview components, add styled-components theme to project based on chakra ui library

PedroEdu6786 committed:
:white_check_mark: feat: add test for header, previewlist and homepage components

Add test coverage report in coverage.txt, implemented tests for rendering

PedroEdu6786 committed:
:lipstick: fix: background color on VideoPreview component

fixed background color change on some VideoPreview cards

PedroEdu6786 committed:
:art: feat: add code review refactoring PR 1

Extracted map function into outside renderVideoItem function

PedroEdu6786 committed:
:sparkles: feat: add useYoutubeApi hook, dynamic videoDetails page

Add dynamic links to videodetails page, add new hook for youtube's api data fetching, add new axios dependency, change header drawer link to home using react-router, add react-router navigation, refactor styles

PedroEdu6786 committed:
:sparkles: feat: dynamic videos from youtube api, useContext state

Add useContext state management, useGlobalState hook, fetch videos from youtube api

PedroEdu6786 committed:
:sparkles: feat: optimize home page using globalstate

update youtubeapi hook to update globalstate, homepage now uses globalstate videos instead of useYoutubeApi data

PedroEdu6786 committed:
:white_check_mark: feat: refactor tests

Pass all test for all test files, updated coverage.txt for more details

PedroEdu6786 committed:
:white_check_mark: feat: Add global context to home test

refactor test for homepage and remove unused commented code on videodetails

PedroEdu6786 committed:
:white_check_mark: feat: add youtubeApi hook tests, fix homepage render test warning

Add new useYoutubeApi hook tests, add new dependency for hook testing, fix warnings on homepage tests run

PedroEdu6786 committed:
:recycle: refactor: useGlobalState, GlobalContextProvider

refactor globalStateProvider to videosProvider, created videosProvider folder structure, refactor useGlobalState for reusability for every provider

PedroEdu6786 committed:
:sparkles: feat: add theme toggle, add dark theme

Add ThemeContentProvider for theme global context management, add toggle theme to header

PedroEdu6786 committed:
:white_check_mark: feat: add tests for useYoutubeApi and helpers

update useYoutubeApi hook test with axios mock functions, add test for failing case on useYoutubeApi. Add tests for helper functions (youtubeClient and buildQueryParams)

PedroEdu6786 committed:
:white_check_mark: feat: Add useGlobalState tests

add new useGlobalState tests for: default state, update state and default dispatch case

PedroEdu6786 committed:
:white_check_mark: feat: update homepage test

fix homepage test warnings, add test for data fetching from api

PedroEdu6786 committed:
:white_check_mark: feat: add test for reducers

add test cases for videosReducer and themeReducer

PedroEdu6786 committed:
:white_check_mark: feat: add Header and PreviewRelatedList tests

Add test for theme toggle on header component. Add test for render related video data

PedroEdu6786 committed:
:recycle: feat: add code review changes from 2nd delivery

move header styles to styles file, add unit testing for overlay on header, new test for renderVideoItem

cehawz commented on src/components/Header/Header.component.jsx:
```diff
@@ -0,0 +1,76 @@
+
+        <RightNav>
+          <ThemeButton as="button" data-testid="themeButton" onClick={handleTheme}>
+            {theme.theme === 'light' ? (
```
Esta linea necesita unit test.

PedroEdu6786 replied:
Hay un test que prueba el cambio de iconos para probar la condición

cehawz commented on src/pages/Home/Home.page.jsx:
```diff
@@ -1,38 +1,32 @@
+      {videos && <PreviewList videos={videos} />}
+      {/* Error when data fetching */}
+      {error && <Heading>There was an error</Heading>}
+    </BodyContainer>
```
Falta un unit test para el render de `PreviewList` y `Heading`

PedroEdu6786 replied:
`PreviewList` es testeado en "loads videos from api" con el data-testid "previewList"

cehawz commented on src/utils/helpers.js:
```diff
@@ -0,0 +1,36 @@
+const SNIPPET = 'snippet';
+const VIDEO = 'video';
+
+// youtube api axios config
```
Esto podria estar en un archivo usado exclusivamente para el consumo de servicios del API de YTB.

PedroEdu6786 replied:
No me quedo claro esta parte

cehawz commented on src/pages/Home/Home.page.jsx:
```diff
@@ -1,38 +1,32 @@
-  const sectionRef = useRef(null);
-  const { authenticated, logout } = useAuth();
+  const { error, fetchVideos } = useYoutubeApi();
+  const { globalState } = useContext(VideosContext);
```
Este fetch Videos deberia ser un action. Con la implementación del state management necesitas consultar el API desde los actions.

PedroEdu6786 replied:
No estoy seguro de como llevar ese cambio a cabo, ya que el error y loading state se encuentran en useYoutubeApi hook, y estos states son modificados dentro de la función `fetchVideos` y los hooks solamente pueden ser utilizados en functional react components

cehawz reviewed (changes requested):
Creo que deberias de renombrar los archivos de VideoProvider de la siguiente manera:

- `VideoProvider.jsx` -> `VideoProvider.js`
- `videosReducer.js` -> `VideoProvider.reducer.js`
- `videoTypes.js` -> `VideoProvider.types.js`

Crear nuevo archivo para definir actions: 

- `VideosProvider.actions.js`

En general lo veo bastante bien.

Hay algunos test que fallan:

![image](https://user-images.githubusercontent.com/77181879/130146915-5991068f-2440-4099-b0ea-d6e4d1557fff.png)

cehawz commented:
Veo bastante bien tu codigo, solo te hice unas recomendaciones. Me avisas para darle otro review.

PedroEdu6786 committed:
:recycle: feat: VideoProvider and ThemeContextProvider

Change file name extension to keep consistency, add actions file for each provider, add changes for files affected

PedroEdu6786 committed:
:white_check_mark: feat: add test for homepage error heading

Add render test for error heading when api request fails

PedroEdu6786 commented:
@cehawz Parece ser que le error de testing tiene que ver más con codesandbox que con mi código, al parecer hay errores al utilizar jest.mock() y por eso dentro de codesandbox falla mi mock de youtubeClient, sin embargo ya probé en 2 distintos ordenadores de forma local y no parece repetirse el error que aparece en codesandbox.

Me gustaría invitarlo a clonar el branch del proyecto para realizar las pruebas pertinentes y verificar su funcionalidad.

PedroEdu6786 committed:
:recycle: feat: youtubeHelpers file

Move youtubeClient to youtubeHelpers file, change files affected by refactor
